{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CREATING the Database and user\n",
    "\n",
    "## Getting Started: Don't run this code in anaconda!\n",
    "\n",
    "First, start by following the steps here up to `Create a Database`: https://medium.com/@FranckPachot/postgresql-and-jupyter-notebook-e7b68cb6427d\n",
    "\n",
    "Then, open your Anaconda Prompt (on Windows) or your Terminal (on Mac) and run the following:\n",
    "\n",
    "## For windows users\n",
    "\n",
    "Make sure to sub your `USERNAME` on your machine here. \n",
    "\n",
    "```\n",
    "#set the database location\n",
    "#add PGDATA to your environment variables\n",
    "set PGDATA=C:\\Anaconda\\pgdata\n",
    "mkdir %PGDATA%\n",
    "pg_ctl initdb\n",
    "pg_ctl start\n",
    "createdb <USERNAME>\n",
    "psql\n",
    "```\n",
    "\n",
    "## For Mac users\n",
    "\n",
    "Make sure to sub your `USERNAME` on your machine here. \n",
    "\n",
    "```\n",
    "#make sure to add this export line to your .bashrc!\n",
    "#set the database location\n",
    "export PGDATA=~/pgdata\n",
    "mkdir $PGDATA\n",
    "pg_ctl initdb\n",
    "pg_ctl start\n",
    "createdb <USERNAME>\n",
    "#open the postgres prompt\n",
    "psql \n",
    "```\n",
    "\n",
    "## For Both types of USERS\n",
    "\n",
    "in the PSQL prompt, type in the following:\n",
    "\n",
    "```\n",
    "CREATE DATABASE ensembl\n",
    "exit\n",
    "```\n",
    "\n",
    "Then continue on here, running the cells below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading Data Into Our Database\n",
    "\n",
    "The first thing we do is to load the `sql` extension, which enables us to run SQL statements directly (make sure that that you have the `ipython-sql` package installed using `conda install -y -c conda-forge ipython-sql` from the Anaconda prompt. Once that is loaded, we connect to our database.\n",
    "\n",
    "Remember to sub your username for `<USERNAME>` here!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: postgres@ensembl'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%reload_ext sql\n",
    "\n",
    "##Connect to the database\n",
    "#%sql postgresql://postgres:postpost@localhost:5433/ensembl\n",
    "%sql postgresql://<USERNAME>@localhost/ensembl            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://postgres:***@localhost:5433/ensembl\n",
      "1 rows affected.\n",
      "+------------------------------------------------------------+\n",
      "|                          version                           |\n",
      "+------------------------------------------------------------+\n",
      "| PostgreSQL 12.1, compiled by Visual C++ build 1914, 64-bit |\n",
      "+------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "pg_version=%sql select version()\n",
    "print(pg_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating the Database Tables\n",
    "\n",
    "Here we load our data into our database. We first need to specify the data types used in our table. Postgres has a lot of different data types.\n",
    "\n",
    "## Before you run this\n",
    "\n",
    "Note that Postgres requires the **absolute** file path to the data files. So you will need to modify the paths that are after the `FROM` statements to be the absolute paths of where your datafiles are. So adjust these paths to fit your machine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://postgres:***@localhost:5433/ensembl\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "168617 rows affected.\n",
      "Done.\n",
      "168617 rows affected.\n",
      "Done.\n",
      "22799 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql \n",
    "drop table if exists gene;\n",
    "drop table if exists transcript;\n",
    "drop table if exists gene2transcript;\n",
    "\n",
    "CREATE TABLE gene2transcript\n",
    "(\n",
    "    ensembl_gene_id character(25),\n",
    "    ensembl_transcript_id character(25)\n",
    ");\n",
    "COPY gene2transcript(ensembl_gene_id, ensembl_transcript_id)\n",
    "FROM 'c:/Code/BMI535slides/data/ensembl_gene_transcript.csv' DELIMITER ',' CSV HEADER;\n",
    "\n",
    "CREATE TABLE transcript\n",
    "  (\n",
    "     ensembl_transcript_id character(25),\n",
    "      transcript_start integer,\n",
    "      transcript_end integer,\n",
    "      transcript_type character varying\n",
    "  );\n",
    "    \n",
    "COPY transcript(ensembl_transcript_id, transcript_start, transcript_end, transcript_type)\n",
    "FROM 'c:/Code/BMI535slides/data/ensembl_transcript.csv' DELIMITER ',' CSV HEADER;\n",
    "\n",
    "CREATE TABLE gene\n",
    "  (\n",
    "      ensembl_gene_id character(25),\n",
    "      gene_strand integer,\n",
    "      gene_end integer,\n",
    "      gene_start integer,\n",
    "      chromosome character varying,\n",
    "      gene_symbol character varying\n",
    "  );\n",
    "    \n",
    "COPY gene(ensembl_gene_id, gene_strand, gene_end, gene_start, chromosome, gene_symbol) \n",
    "FROM 'c:/Code/BMI535slides/data/ensembl_gene.csv' DELIMITER ',' CSV HEADER;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ensuring we have loaded our data correctly\n",
    "\n",
    "Now we're going to run a couple of SQL commands to ensure that we've loaded our data in correctly. Use this page as a reference when you are doing the exercises."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://postgres:***@localhost:5433/ensembl\n",
      "10 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>ensembl_gene_id</th>\n",
       "        <th>gene_strand</th>\n",
       "        <th>gene_end</th>\n",
       "        <th>gene_start</th>\n",
       "        <th>chromosome</th>\n",
       "        <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198888          </td>\n",
       "        <td>1</td>\n",
       "        <td>4262</td>\n",
       "        <td>3307</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198763          </td>\n",
       "        <td>1</td>\n",
       "        <td>5511</td>\n",
       "        <td>4470</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-ND2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198804          </td>\n",
       "        <td>1</td>\n",
       "        <td>7445</td>\n",
       "        <td>5904</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-CO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198712          </td>\n",
       "        <td>1</td>\n",
       "        <td>8269</td>\n",
       "        <td>7586</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-CO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000228253          </td>\n",
       "        <td>1</td>\n",
       "        <td>8572</td>\n",
       "        <td>8366</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-ATP8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198899          </td>\n",
       "        <td>1</td>\n",
       "        <td>9207</td>\n",
       "        <td>8527</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-ATP6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198938          </td>\n",
       "        <td>1</td>\n",
       "        <td>9990</td>\n",
       "        <td>9207</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-CO3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198840          </td>\n",
       "        <td>1</td>\n",
       "        <td>10404</td>\n",
       "        <td>10059</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-ND3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000212907          </td>\n",
       "        <td>1</td>\n",
       "        <td>10766</td>\n",
       "        <td>10470</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-ND4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198886          </td>\n",
       "        <td>1</td>\n",
       "        <td>12137</td>\n",
       "        <td>10760</td>\n",
       "        <td>MT</td>\n",
       "        <td>MT-ND4</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('ENSG00000198888          ', 1, 4262, 3307, 'MT', 'MT-ND1'),\n",
       " ('ENSG00000198763          ', 1, 5511, 4470, 'MT', 'MT-ND2'),\n",
       " ('ENSG00000198804          ', 1, 7445, 5904, 'MT', 'MT-CO1'),\n",
       " ('ENSG00000198712          ', 1, 8269, 7586, 'MT', 'MT-CO2'),\n",
       " ('ENSG00000228253          ', 1, 8572, 8366, 'MT', 'MT-ATP8'),\n",
       " ('ENSG00000198899          ', 1, 9207, 8527, 'MT', 'MT-ATP6'),\n",
       " ('ENSG00000198938          ', 1, 9990, 9207, 'MT', 'MT-CO3'),\n",
       " ('ENSG00000198840          ', 1, 10404, 10059, 'MT', 'MT-ND3'),\n",
       " ('ENSG00000212907          ', 1, 10766, 10470, 'MT', 'MT-ND4L'),\n",
       " ('ENSG00000198886          ', 1, 12137, 10760, 'MT', 'MT-ND4')]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * from gene LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://postgres:***@localhost:5433/ensembl\n",
      "10 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>ensembl_transcript_id</th>\n",
       "        <th>transcript_start</th>\n",
       "        <th>transcript_end</th>\n",
       "        <th>transcript_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361390          </td>\n",
       "        <td>3307</td>\n",
       "        <td>4262</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361453          </td>\n",
       "        <td>4470</td>\n",
       "        <td>5511</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361624          </td>\n",
       "        <td>5904</td>\n",
       "        <td>7445</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361739          </td>\n",
       "        <td>7586</td>\n",
       "        <td>8269</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361851          </td>\n",
       "        <td>8366</td>\n",
       "        <td>8572</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361899          </td>\n",
       "        <td>8527</td>\n",
       "        <td>9207</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000362079          </td>\n",
       "        <td>9207</td>\n",
       "        <td>9990</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361227          </td>\n",
       "        <td>10059</td>\n",
       "        <td>10404</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361335          </td>\n",
       "        <td>10470</td>\n",
       "        <td>10766</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENST00000361381          </td>\n",
       "        <td>10760</td>\n",
       "        <td>12137</td>\n",
       "        <td>protein_coding</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('ENST00000361390          ', 3307, 4262, 'protein_coding'),\n",
       " ('ENST00000361453          ', 4470, 5511, 'protein_coding'),\n",
       " ('ENST00000361624          ', 5904, 7445, 'protein_coding'),\n",
       " ('ENST00000361739          ', 7586, 8269, 'protein_coding'),\n",
       " ('ENST00000361851          ', 8366, 8572, 'protein_coding'),\n",
       " ('ENST00000361899          ', 8527, 9207, 'protein_coding'),\n",
       " ('ENST00000362079          ', 9207, 9990, 'protein_coding'),\n",
       " ('ENST00000361227          ', 10059, 10404, 'protein_coding'),\n",
       " ('ENST00000361335          ', 10470, 10766, 'protein_coding'),\n",
       " ('ENST00000361381          ', 10760, 12137, 'protein_coding')]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM transcript LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://postgres:***@localhost:5433/ensembl\n",
      "10 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>ensembl_gene_id</th>\n",
       "        <th>ensembl_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198888          </td>\n",
       "        <td>ENST00000361390          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198763          </td>\n",
       "        <td>ENST00000361453          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198804          </td>\n",
       "        <td>ENST00000361624          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198712          </td>\n",
       "        <td>ENST00000361739          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000228253          </td>\n",
       "        <td>ENST00000361851          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198899          </td>\n",
       "        <td>ENST00000361899          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198938          </td>\n",
       "        <td>ENST00000362079          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198840          </td>\n",
       "        <td>ENST00000361227          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000212907          </td>\n",
       "        <td>ENST00000361335          </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ENSG00000198886          </td>\n",
       "        <td>ENST00000361381          </td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('ENSG00000198888          ', 'ENST00000361390          '),\n",
       " ('ENSG00000198763          ', 'ENST00000361453          '),\n",
       " ('ENSG00000198804          ', 'ENST00000361624          '),\n",
       " ('ENSG00000198712          ', 'ENST00000361739          '),\n",
       " ('ENSG00000228253          ', 'ENST00000361851          '),\n",
       " ('ENSG00000198899          ', 'ENST00000361899          '),\n",
       " ('ENSG00000198938          ', 'ENST00000362079          '),\n",
       " ('ENSG00000198840          ', 'ENST00000361227          '),\n",
       " ('ENSG00000212907          ', 'ENST00000361335          '),\n",
       " ('ENSG00000198886          ', 'ENST00000361381          ')]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM gene2transcript LIMIT 10;"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
