{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to postgres SQL\n",
    "\n",
    "This notebook is meant to be a short introduction to relational database queries using SQL.\n",
    "\n",
    "Note that you will need to load the database first (see that notebook for more information).\n",
    "\n",
    "Make sure that you change <USERNAME> below to your own username!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load the ipython-sql magic extension\n",
    "#allows us to run SQL directly in notebook cells\n",
    "%reload_ext sql\n",
    "\n",
    "##Connect to the database\n",
    "%sql postgresql://postgres:postpost@localhost:5433/ensembl\n",
    "#%sql postgresql://<USERNAME>@localhost/ensembl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Review: What is a Relational Database?\n",
    "\n",
    "- Collection of tables with *relationships* between them\n",
    "     - We will define this as we go on\n",
    "- Managed by a database management system (DBMS) which controls:\n",
    "     - Who can access which table (role-based access)\n",
    "     - Who can modify which table using *transactions*\n",
    "     - Define how transactions modify the tables\n",
    "     - Provide a way to \"undo\" transactions (rolling back)\n",
    "     - Back up databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What is Structured Query Language?\n",
    "\n",
    "- [Standard computer language](https://www.techopedia.com/definition/1245/structured-query-language-sql) for Databases and DBMSs\n",
    "- SQL is used to \n",
    "    - query data\n",
    "    - insert data\n",
    "    - update data\n",
    "    - modify data\n",
    "- *Unfortunately*, SQL is defined differently for different DBMSs\n",
    "    - [Many \"dialects\" of SQL](https://en.wikibooks.org/wiki/SQL_Dialects_Reference/Introduction) (MS SQL, MySQL, Postgres SQL, SQLite, etc)\n",
    "    - This means that different DBMSs have different capabilities\n",
    "        - SQLite: very lightweight and minimal implementation \n",
    "        - PostgreSQL: much more heavyweight\n",
    "        - MS SQL Server: still used quite a bit\n",
    "\n",
    "More on the differences between DBMSs here: http://troels.arvin.dk/db/rdbms/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SELECTing FROM a table in the database\n",
    "\n",
    "All SQL queries have the following form:\n",
    "\n",
    "```\n",
    "SELECT *columns*  <- required\n",
    "FROM *table*      <- required\n",
    "WHERE *criteria*  <- this clause is optional  \n",
    "```\n",
    "\n",
    "The very first thing we can do is return all the columns from a query using `*` (the SQL wildcard character). But remember, we need a `FROM` clause as well to make our query complete."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene LIMIT 20;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WHERE\n",
    "\n",
    "WHERE is an optional clause, but it lets us add filtering criteria to your query. We use the LIMIT clause to only show the first 20 lines of our table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene WHERE chromosome = '14' LIMIT 20;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Boolean Operations\n",
    "\n",
    "We can chain criteria together by using the `AND`/`OR` boolean operations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "SELECT * FROM gene WHERE gene_end < 100000 AND chromosome = '10'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Think about it\n",
    "\n",
    "Should an AND query be *larger* or *smaller* than an OR query?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "SELECT * FROM gene WHERE gene_end < 100000 OR chromosome = '10'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SELECTING columns\n",
    "\n",
    "You can select individual columns of each table using the SELECT statement. \n",
    "\n",
    "Note that we have to put our chromosome (`14`) in quotes since the datatype of the column is character."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql \n",
    "SELECT ensembl_gene_id, gene_start, chromosome \n",
    "FROM gene\n",
    "WHERE chromosome = '14'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Your Turn\n",
    "\n",
    "Write a SELECT statement that returns everything from the `gene` table if gene_symbol = `FGFR2`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aliases using AS\n",
    "\n",
    "One useful trick we can use are *aliases*. This becomes more important as we join tables together, since it:\n",
    "\n",
    "1. Saves us typing\n",
    "2. Makes our query more clear and specific. \n",
    "\n",
    "You can see that `ensembl_gene_id` might be in multiple tables. Postgres gets confused if we don't.\n",
    "\n",
    "In general, you can use an alias *before* you define it in the query, which is confusing. You just need to define it somewhere (usually in the FROM clause).\n",
    "\n",
    "When you refer to a column, it is best to preface it with the *alias* so your query is exact."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql \n",
    "SELECT g.ensembl_gene_id, g.gene_start, g.chromosome \n",
    "    FROM gene as g\n",
    "    WHERE chromosome = '14' LIMIT 10;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also rename our columns using AS:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql \n",
    "SELECT ensembl_gene_id as ensembl, gene_start, chromosome FROM gene\n",
    "WHERE chromosome = '14' LIMIT 20;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aggregating using COUNT\n",
    "\n",
    "The COUNT verb in SQL lets us count things. If we use COUNT(ensembl_gene_id), it will count the number of "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(ensembl_transcript_id) FROM transcript;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(DISTINCT ensembl_gene_id) FROM gene2transcript;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GROUP BY\n",
    "\n",
    "GROUP BY is extremely useful if we want to produce a table of counts. Note that to produce a table with counts, we need to return both `chromosome` and `COUNT(chromosome)`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT chromosome, COUNT(chromosome) FROM gene GROUP BY chromosome;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ORDER BY\n",
    "\n",
    "That's great and everything, but that doesn't answer the question of which chromosome has the largest number of mapped genes. \n",
    "\n",
    "By adding an `ORDER BY` clause followed by a `DESC`, we can sort our result table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql \n",
    "    SELECT chromosome, COUNT(chromosome) as count \n",
    "        FROM gene \n",
    "        GROUP BY chromosome \n",
    "        ORDER BY count DESC;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Your Turn\n",
    "\n",
    "How many genes are mapped to the `1` (+) strand and how many genes are mapped to the `-1` (-) strand?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# JOINing the Tables\n",
    "\n",
    "Of course, our `transcript` and `gene` tables aren't that useful by themselves. We need to integrate information in these tables to produce useful queries.\n",
    "\n",
    "We will use what are called JOINs on the data, to produce a table that has information about genes and their transcripts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The Different JOIN types\n",
    "\n",
    "I apologize for the different tables here. However, these figures took a long time to make and I don't have the time to make bioinformatics specific ones.\n",
    "\n",
    "- INNER JOIN\n",
    "\n",
    "Only retains the rows which are in common between the two joined tables.\n",
    "\n",
    "![](docs/image/Slide3.JPG)\n",
    "\n",
    "- LEFT JOIN\n",
    "\n",
    "Retains everything in INNER JOIN, plus those rows in the LEFT table.\n",
    "\n",
    "![](docs/image/Slide4.JPG)\n",
    "\n",
    "- RIGHT JOIN\n",
    "\n",
    "Retains everything in INNER JOIN, plus those rows in the RIGHT table.\n",
    "\n",
    "- OUTER JOIN\n",
    "\n",
    "Retains ALL rows in both tables, regardless of whether there is matching criteria."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quiz Yourself\n",
    "\n",
    "How many rows would there be in the results table above if we RIGHT JOINed instead?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LEFT JOIN in action\n",
    "\n",
    "Now, let's try a `LEFT JOIN` on our data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql \n",
    "    SELECT g.ensembl_gene_id, \n",
    "        g.gene_start, \n",
    "        g.gene_symbol, \n",
    "        t2g.ensembl_transcript_id\n",
    "    FROM gene AS g\n",
    "    LEFT JOIN gene2transcript AS t2g \n",
    "        ON g.ensembl_gene_id = t2g.ensembl_gene_id \n",
    "    LIMIT 20;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You may recall that we have **three** tables: `gene`, `gene2transcript` and `transcript`. To join together `gene` and `transcript`, our `LEFT JOIN` has to include all three tables. \n",
    "\n",
    "We can do this by adding another `LEFT JOIN` clause to our query. Here we're adding in the `transcript` table by including the criteria `t2g.ensembl_transcript_id` = `t.ensembl_transcript_id`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql \n",
    "    SELECT g.ensembl_gene_id, g.gene_start, g.gene_symbol, t.ensembl_transcript_id, t.transcript_start\n",
    "    FROM gene AS g\n",
    "    LEFT JOIN gene2transcript AS t2g \n",
    "        ON g.ensembl_gene_id = t2g.ensembl_gene_id \n",
    "    LEFT JOIN transcript AS t \n",
    "        ON t2g.ensembl_transcript_id = t.ensembl_transcript_id\n",
    "    LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use Case: Finding Overlapping Genes\n",
    "\n",
    "One query we might do is to find all of the overlapping genes in our database. \n",
    "\n",
    "That is, we want those genes where \n",
    "\n",
    "```\n",
    "gene1.gene_symbol != gene2.gene_symbol\n",
    "gene1.chromosome = gene2.chromosome\n",
    "gene1.start > gene2.end and \n",
    "gene2.end < gene.1 start and\n",
    "```\n",
    "\n",
    "How do we do this, since we only have 1 table? We need to do what is called a SELF JOIN. \n",
    "\n",
    "Basically, we make two aliases for our table, called `g1` and `g2`. Then we can join these two tables just like any other join. Here we use an `INNER JOIN`.\n",
    "\n",
    "Note that additional filtering criteria are added in our `WHERE` clause. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "    SELECT g1.gene_symbol AS gene1, \n",
    "        g2.gene_symbol AS gene2,\n",
    "        g1.gene_start as g1_start,\n",
    "        g2.gene_start as g2_start,\n",
    "        g1.gene_end as g1_end,\n",
    "        g2.gene_end as g2_end\n",
    "    FROM gene AS g1\n",
    "    INNER JOIN gene AS g2\n",
    "        ON g1.gene_symbol != g2.gene_symbol \n",
    "    WHERE g1.gene_start <= g2.gene_end\n",
    "        AND g1.gene_end >= g2.gene_start\n",
    "        AND g1.chromosome = g2.chromosome\n",
    "        LIMIT 20;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using EXPLAIN ANALYZE to find out why our query is slow\n",
    "\n",
    "Because we are joining and scanning many tables for a query, some of the more complicated ones may take a lot longer to run.\n",
    "\n",
    "If we add `EXPLAIN ANALYZE` to the beginning of our query, we will understand what tasks the Database Management System (DBMS) take the longest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Need to run this for demo purposes\n",
    "%sql DROP INDEX IF EXISTS ChrStartEnd;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "    EXPLAIN ANALYZE\n",
    "    SELECT g1.gene_symbol AS gene1, \n",
    "        g2.gene_symbol AS gene2,\n",
    "        g1.gene_start as g1_start,\n",
    "        g2.gene_start as g2_start,\n",
    "        g1.gene_end as g1_end,\n",
    "        g2.gene_end as g2_end\n",
    "    FROM gene AS g1\n",
    "    INNER JOIN gene AS g2\n",
    "        ON g1.gene_symbol != g2.gene_symbol \n",
    "    WHERE g1.gene_start <= g2.gene_end\n",
    "        AND g1.gene_end >= g2.gene_start\n",
    "        AND g1.chromosome = g2.chromosome;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Indexing: making a query faster\n",
    "\n",
    "Querying takes time, because we have to scan the whole table.\n",
    "\n",
    "However, if there is a column we query a lot, we can create a **index** for it. This index allows us to query the table faster.\n",
    "\n",
    "Much like the other data structures we use in Python, an *index* takes the form of a B-Tree, which allows for much more rapid searching of the data. \n",
    "\n",
    "Why don't we index everything? Well, indexes take a lot of disk space. So we are better off picking and choosing which columns we want to index. This is based on what kinds of queries and searches that are the most commonly done on a table.\n",
    "\n",
    "There are data engineers whose job it is to tune the database to make queries run faster. Don't worry too much if this is confusing right now. Just know who you're going to talk with when you need a database query to run lightning fast."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "CREATE INDEX ChrStartEnd on gene(chromosome, gene_start, gene_end)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now try running the query again with `EXPLAIN ANALYZE`. How much faster was it on your machine?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "    EXPLAIN ANALYZE\n",
    "    SELECT g1.gene_symbol AS gene1, \n",
    "        g2.gene_symbol AS gene2,\n",
    "        g1.gene_start as g1_start,\n",
    "        g2.gene_start as g2_start,\n",
    "        g1.gene_end as g1_end,\n",
    "        g2.gene_end as g2_end\n",
    "    FROM gene AS g1\n",
    "    INNER JOIN gene AS g2\n",
    "        ON g1.gene_symbol != g2.gene_symbol \n",
    "    WHERE g1.gene_start <= g2.gene_end\n",
    "        AND g1.gene_end >= g2.gene_start\n",
    "        AND g1.chromosome = g2.chromosome;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Indexing and Primary Keys\n",
    "\n",
    "In our next session, we will be learning some useful terminology for our database: \n",
    "- *Primary Keys*, which uniquely identify a row in a table\n",
    "- *Foreign Keys*, which refer to a key in another table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Useful to know: Subqueries\n",
    "\n",
    "This has been a lot of SQL to throw at you! So I'm making this last section optional.\n",
    "\n",
    "Sometimes it's useful to break up complicated queries into subqueries. Basically any query that returns a table can be used in a FROM statement, but you must create an alias for the subquery."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "SELECT ensembl_gene_id, gene_symbol\n",
    "    FROM\n",
    "    (SELECT * FROM gene WHERE chromosome = '1') AS subquery\n",
    "    WHERE subquery.gene_start >= 1000000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another clause is `UNION`, which can combine two subqueries, basically paste one of the subquery tables on top of another."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%sql\n",
    "    (SELECT * FROM gene WHERE gene_symbol = 'TP53') \n",
    "    UNION\n",
    "    (SELECT* FROM gene WHERE gene_symbol = 'FGFR2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fix Me!\n",
    "\n",
    "Why does this UNION query not work? \n",
    "\n",
    "Modify the bottom subquery to work with the top subquery."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "%%sql\n",
    "    (SELECT ensembl_gene_id, gene_symbol \n",
    "     FROM gene \n",
    "     WHERE gene_symbol = 'TP53')\n",
    "    UNION\n",
    "    (SELECT *\n",
    "     FROM gene \n",
    "     WHERE gene_symbol = 'FGFR2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What's Next?\n",
    "\n",
    "We'll learn some of the intricacies about Database Design and Normalization. Stay tuned!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Acknowledgements\n",
    "\n",
    "This notebook was adapted from a tutorial by Christina Zheng."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
