<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Intro to Big Data and Command Line Review</title>
    <meta charset="utf-8" />
    <meta name="author" content="Ted Laderas" />
    <meta name="date" content="2021-01-05" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Intro to Big Data and Command Line Review
## BMI 535/635
### Ted Laderas
### 2021-01-05

---





---

class: center, middle
# Welcome and Introduction

---
class:center, middle
# This is a Safe Space to Ask Questions

---
# Code of Conduct

This class is governed by the [BioData Club Code of Conduct](https://biodata-club.github.io/code_of_conduct/).

This class is meant to be a psychologically safe space where it's ok to ask questions.

I want to normalize your own curiosity and fuel your desire to learn more.

If you are disruptive to class learning or disparaging to other students, I may mute you for the day.

---
# Learning Objectives

1. **Utilize** alternative data storage and data structures to handle Big Datasets
1. **Create**, **Normalize**, **Load** and **Query** Relational Databases
1. **Understand** the types of NoSQL Databases and their uses
1. **Create** and **Run** jobs on High Performance Computing systems including Exacloud and Amazon Web Services
1. **Understand** and **utilize** the map-reduce programming pattern
1. **Reproduce**, **Implement**, **Debug**, and **Document** an established biomedical data workflow
1. **Utilize** container-based workflows on a remote systems
1. (BMI 635 only) **Design** and **Evaluate** the efficiency of parallelized analysis workflows

---

# Structure of this Course

- Command Line Review
- Useful Data Structures in R/Python for handling Big Data
- High Performance Computing (HPC)
  - Exacloud and Amazon Web Services (AWS)
- Workflows
  - Docker/GATK 
- Databases
  - Relational Databases
  - NoSQL Databases
- Final Project
---

# Grading

- Assignments: 40%
- Midterm Exam: 20%
- Final Project: 40%

---

# Office Hours/Appointments

Please contact us if you need to make an appointment

- laderast@ohsu.edu
- mooneymi@ohsu.edu

---

# BMI 535 Slack

https://bmi535.slack.com

Invitations will be sent by email.

Feel free to ask questions and answer them.

---
class: center, middle

# Why Big Data?

---
&lt;img src="image/5V.png" width="760" /&gt;
---
# Big Data and Personalized Medicine

&lt;img src="image/health_data_types.png" width="1267" /&gt;

---
# Precision Medicine

&lt;img src="image/precision.png" width="1076" /&gt;
---
# We need lots of genomes to make this happen

.pull-left[

Need to identify associations with disease

- Phenotype information
- Genotype Information]

.pull-right[

  &lt;img src="image/gwas_infographic.jpg"&gt;
]

---
# 1000 Genomes Project

International project to construct a foundational data set of human genetics

- discover virtually all common human variation
	- single nucleotide polymorphisms
	- structural variants
	- haplotypes

- develop sequence analysis methods, tools, and reagents that can be transferred to other sequencing projects

---
# (More than) 1000 genomes:

2504 individuals, 26 populations

&lt;img src="image/1000genomes.png" width="1244" /&gt;
---
# 1000 genomes information

2504 individuals, 26 populations
 - low-coverage whole genome sequencing (mean of 7.4X)
 - deep exome sequencing (mean of 65.7X)
 - dense microarray genotyping

84.7M SNPs, 3.6 million indels, 60,000 structural variants
 - &gt;99% of SNP variants &gt;1% frequency (what does that mean?)

---

# Variation in 1000 genomes

&lt;img src="image/1000var.png" width="941" /&gt;
---
class: center, middle

# We need more genomes!

---
# Open Respositories of Genomic Data

&lt;img src="image/pgp.png" width="1280" /&gt;

---
# GA4GH

&lt;img src="image/ga4gh.png" width="1267" /&gt;

---

# DNA: What are some issues?

- Privacy/identifiability
- Using specific genomic markers and de-identified information, we can uniquely identify about 87% of the US population that has participated in either 1000 genomes or GA4GH
---

# Need to protect privacy 

Controlled Access
  - approved access
  - recording/audited queries
  - participant based access

---
# NIH Policy

&lt;img src="image/nih.png" width="1243" /&gt;

---
class: center, middle

# Back to Big Data

---
# Big Data Processing Strategies

- Massively Parallel Processing
- Cluster Computing
- Map/Reduce

---
# Storing and Querying Big Data

- Data Structures
- Relational Databases
- NoSQL (Not only SQL) Databases
  - Document stores
  - Column stores
  - Graph Databases

[Modern Data Formats for Big Bioinformatics Data
Analytics](https://arxiv.org/ftp/arxiv/papers/1707/1707.05364.pdf)
  
---
class: center, middle

# Command Line Review 

---
# Our Safe Space: state.ohsu.edu

---
# Reference: Linux Fundamentals

- Paul Cobbaut (see course materials week 1)

---
# Review: Files

- 9.2: Remember that all files are case-sensitive
- 9.3: Everything in unix is a file
    - including folders and devices!

---
# Directory Structure

`/home/its/ug1/eev51vn/report.doc`

&lt;img src="image/file_directories.png" width="695" /&gt;

---
# Relative vs. Absolute Paths (8.3)

If I'm in `/home/its/ug1`, how can I refer to `report.doc` from this folder?

&lt;img src="image/file_directories.png" width="695" /&gt;

---
# Useful Commands for manipulating files (Chapters 8-10)

Be familiar with the following commands: 

- `ls` - listing files within a directory
- `mv` - move/rename a file
- `rm` - removing files 
- `rm -r` - removing files within a directory (recursive)
- `grep` - look for a pattern/regular expression in a file
- `chmod` - file permissions
- `pwd` - current directory
- `head/cat/tail` - look at the contents of a file

---
# Learn a Command Line text editor (vi/nano, Chapter 22)

- `vi` - guaranteed to be on any linux system
  - A little harder to learn
  - command mode versus insert mode
  - extremely powerful for searching/replacing
- `nano` - usually on any linux system
  - A little easier to learn

---
# Scripting Configuration (chapter 14)

- `$PATH` variable (where applications live, 14.8)
- Environment variables (passwords and connections, 14.9)
  - using `.bash_profile` or `.bashrc` files to specify environment variables using `export` (14.10)

---
# Let's Get onto State

https://github.com/laderast/BMI535slides/blob/master/00_unix_basics.md

---
# Suggested Reading

[The Missing Semester of Your CS Education](https://missing.csail.mit.edu/)

Specifically:

- [Course Overview and the Shell](https://missing.csail.mit.edu/2020/course-shell/)
- [Command Line Environment](https://missing.csail.mit.edu/2020/command-line/)
- [Shell Scripting](https://missing.csail.mit.edu/2020/shell-tools/)

---

# Optional Reading

- [Editors: Vim](https://missing.csail.mit.edu/2020/editors/)
- [Version Control: Git](https://missing.csail.mit.edu/2020/version-control/)

---
# Shell Scripting (Chapters 23-26)

- How to set up a shell script
- What a she-bang is (23.3)

---
# Pipes (Chapter 18)

- Know what a pipe `|` is and why it's useful

---
# Regular Expressions (Chapter 21)

- Used to specify string matching queries
- Highly flexible
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
